File: concepts/01-fundamental-concepts/node/load-balancer.js
/**
 * Load Balancer Implementation
 * Demonstrates different load balancing algorithms
 */

class LoadBalancer {
  constructor(servers, algorithm = 'round-robin') {
    this.servers = servers;
    this.algorithm = algorithm;
    this.currentIndex = 0;
    this.serverConnections = new Map();
    this.serverWeights = new Map();
    
    // Initialize connection counts
    servers.forEach(server => {
      this.serverConnections.set(server, 0);
      this.serverWeights.set(server, 1); // Default weight
    });
  }

  /**
   * Round Robin Algorithm
   * Distributes requests sequentially
   */
  roundRobin() {
    const server = this.servers[this.currentIndex];
    this.currentIndex = (this.currentIndex + 1) % this.servers.length;
    return server;
  }

  /**
   * Least Connections Algorithm
   * Routes to server with fewest active connections
   */
  leastConnections() {
    let minConnections = Infinity;
    let selectedServer = this.servers[0];

    for (const server of this.servers) {
      const connections = this.serverConnections.get(server);
      if (connections < minConnections) {
        minConnections = connections;
        selectedServer = server;
      }
    }

    return selectedServer;
  }

  /**
   * Weighted Round Robin
   * Routes based on server weights
   */
  weightedRoundRobin() {
    const totalWeight = Array.from(this.serverWeights.values())
      .reduce((sum, weight) => sum + weight, 0);
    
    let currentWeight = 0;
    let selectedServer = this.servers[0];

    for (const server of this.servers) {
      currentWeight += this.serverWeights.get(server);
      if (Math.random() * totalWeight < currentWeight) {
        selectedServer = server;
        break;
      }
    }

    return selectedServer;
  }

  /**
   * IP Hash Algorithm
   * Routes based on client IP hash
   */
  ipHash(clientIp) {
    const hash = this.simpleHash(clientIp);
    const index = hash % this.servers.length;
    return this.servers[index];
  }

  /**
   * Simple hash function
   */
  simpleHash(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32-bit integer
    }
    return Math.abs(hash);
  }

  /**
   * Set server weight
   */
  setWeight(server, weight) {
    this.serverWeights.set(server, weight);
  }

  /**
   * Get next server based on algorithm
   */
  getNextServer(clientIp = null) {
    let server;

    switch (this.algorithm) {
      case 'round-robin':
        server = this.roundRobin();
        break;
      case 'least-connections':
        server = this.leastConnections();
        break;
      case 'weighted-round-robin':
        server = this.weightedRoundRobin();
        break;
      case 'ip-hash':
        if (!clientIp) {
          throw new Error('Client IP required for IP hash algorithm');
        }
        server = this.ipHash(clientIp);
        break;
      default:
        server = this.roundRobin();
    }

    // Increment connection count
    this.serverConnections.set(
      server,
      this.serverConnections.get(server) + 1
    );

    return server;
  }

  /**
   * Release connection from server
   */
  releaseConnection(server) {
    const current = this.serverConnections.get(server);
    if (current > 0) {
      this.serverConnections.set(server, current - 1);
    }
  }

  /**
   * Health check for servers
   */
  async healthCheck(server) {
    try {
      // Simulate health check
      const response = await fetch(`http://${server}/health`);
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  /**
   * Get server statistics
   */
  getStats() {
    const stats = {};
    for (const server of this.servers) {
      stats[server] = {
        connections: this.serverConnections.get(server),
        weight: this.serverWeights.get(server)
      };
    }
    return stats;
  }
}

// Example usage
const servers = ['server1:3000', 'server2:3000', 'server3:3000'];

// Round Robin
const roundRobinLB = new LoadBalancer(servers, 'round-robin');
console.log('Round Robin:');
for (let i = 0; i < 5; i++) {
  console.log(`Request ${i + 1}: ${roundRobinLB.getNextServer()}`);
}

// Least Connections
const leastConnectionsLB = new LoadBalancer(servers, 'least-connections');
console.log('\nLeast Connections:');
for (let i = 0; i < 5; i++) {
  const server = leastConnectionsLB.getNextServer();
  console.log(`Request ${i + 1}: ${server}`);
  if (i === 2) {
    leastConnectionsLB.releaseConnection(server);
  }
}

// IP Hash
const ipHashLB = new LoadBalancer(servers, 'ip-hash');
console.log('\nIP Hash:');
const clientIps = ['192.168.1.1', '192.168.1.2', '192.168.1.1'];
clientIps.forEach(ip => {
  console.log(`IP ${ip}: ${ipHashLB.getNextServer(ip)}`);
});

// Weighted Round Robin
const weightedLB = new LoadBalancer(servers, 'weighted-round-robin');
weightedLB.setWeight('server1:3000', 3);
weightedLB.setWeight('server2:3000', 2);
weightedLB.setWeight('server3:3000', 1);
console.log('\nWeighted Round Robin:');
const distribution = {};
for (let i = 0; i < 100; i++) {
  const server = weightedLB.getNextServer();
  distribution[server] = (distribution[server] || 0) + 1;
}
console.log('Distribution after 100 requests:', distribution);

module.exports = LoadBalancer;

File: concepts/01-fundamental-concepts/python/load_balancer.py
"""
Load Balancer Implementation
Demonstrates different load balancing algorithms
"""
from collections import defaultdict
from typing import List, Dict, Optional
import hashlib
import random


class LoadBalancer:
    """Load balancer with multiple algorithms"""
    
    def __init__(self, servers: List[str], algorithm: str = 'round-robin'):
        self.servers = servers
        self.algorithm = algorithm
        self.current_index = 0
        self.server_connections: Dict[str, int] = defaultdict(int)
        self.server_weights: Dict[str, int] = {}
        
        # Initialize weights
        for server in servers:
            self.server_weights[server] = 1
    
    def round_robin(self) -> str:
        """Round Robin Algorithm - distributes requests sequentially"""
        server = self.servers[self.current_index]
        self.current_index = (self.current_index + 1) % len(self.servers)
        return server
    
    def least_connections(self) -> str:
        """Least Connections - routes to server with fewest connections"""
        min_connections = float('inf')
        selected_server = self.servers[0]
        
        for server in self.servers:
            connections = self.server_connections[server]
            if connections < min_connections:
                min_connections = connections
                selected_server = server
        
        return selected_server
    
    def weighted_round_robin(self) -> str:
        """Weighted Round Robin - routes based on server weights"""
        total_weight = sum(self.server_weights.values())
        current_weight = 0
        selected_server = self.servers[0]
        
        random_value = random.random() * total_weight
        
        for server in self.servers:
            current_weight += self.server_weights[server]
            if random_value < current_weight:
                selected_server = server
                break
        
        return selected_server
    
    def ip_hash(self, client_ip: str) -> str:
        """IP Hash - routes based on client IP hash"""
        hash_value = int(hashlib.md5(client_ip.encode()).hexdigest(), 16)
        index = hash_value % len(self.servers)
        return self.servers[index]
    
    def set_weight(self, server: str, weight: int):
        """Set weight for a server"""
        if server in self.servers:
            self.server_weights[server] = weight
        else:
            raise ValueError(f"Server {server} not found")
    
    def get_next_server(self, client_ip: Optional[str] = None) -> str:
        """Get next server based on algorithm"""
        server = None
        
        if self.algorithm == 'round-robin':
            server = self.round_robin()
        elif self.algorithm == 'least-connections':
            server = self.least_connections()
        elif self.algorithm == 'weighted-round-robin':
            server = self.weighted_round_robin()
        elif self.algorithm == 'ip-hash':
            if not client_ip:
                raise ValueError("Client IP required for IP hash algorithm")
            server = self.ip_hash(client_ip)
        else:
            server = self.round_robin()
        
        # Increment connection count
        self.server_connections[server] += 1
        return server
    
    def release_connection(self, server: str):
        """Release connection from server"""
        if self.server_connections[server] > 0:
            self.server_connections[server] -= 1
    
    def get_stats(self) -> Dict[str, Dict]:
        """Get server statistics"""
        stats = {}
        for server in self.servers:
            stats[server] = {
                'connections': self.server_connections[server],
                'weight': self.server_weights[server]
            }
        return stats


def demonstrate_load_balancing():
    """Demonstrate different load balancing algorithms"""
    servers = ['server1:3000', 'server2:3000', 'server3:3000']
    
    # Round Robin
    print("=== Round Robin ===")
    round_robin_lb = LoadBalancer(servers, 'round-robin')
    for i in range(5):
        server = round_robin_lb.get_next_server()
        print(f"Request {i + 1}: {server}")
    
    # Least Connections
    print("\n=== Least Connections ===")
    least_conn_lb = LoadBalancer(servers, 'least-connections')
    for i in range(5):
        server = least_conn_lb.get_next_server()
        print(f"Request {i + 1}: {server}")
        if i == 2:
            least_conn_lb.release_connection(server)
    
    # IP Hash
    print("\n=== IP Hash ===")
    ip_hash_lb = LoadBalancer(servers, 'ip-hash')
    client_ips = ['192.168.1.1', '192.168.1.2', '192.168.1.1']
    for ip in client_ips:
        server = ip_hash_lb.get_next_server(ip)
        print(f"IP {ip}: {server}")
    
    # Weighted Round Robin
    print("\n=== Weighted Round Robin ===")
    weighted_lb = LoadBalancer(servers, 'weighted-round-robin')
    weighted_lb.set_weight('server1:3000', 3)
    weighted_lb.set_weight('server2:3000', 2)
    weighted_lb.set_weight('server3:3000', 1)
    
    distribution = defaultdict(int)
    for _ in range(100):
        server = weighted_lb.get_next_server()
        distribution[server] += 1
    
    print("Distribution after 100 requests:")
    for server, count in distribution.items():
        print(f"  {server}: {count}")


if __name__ == '__main__':
    demonstrate_load_balancing()

